lib/services/role.go

```diff
@@ -3288,14 +3291,32 @@ func (set RoleSet) GetAllowedSearchAsRoles() []string {
 	}
 	for _, role := range set {
 		for _, a := range role.GetSearchAsRoles(types.Allow) {
-			if _, ok := denied[a]; !ok {
+			filterResult := true
+			if len(allowFilters) > 0 {
+				for _, filter := range allowFilters {
+					if !filter(a) {
+						filterResult = false
+						break
+					}
+				}
+			}
+			if _, ok := denied[a]; !ok && filterResult {
 				allowed = append(allowed, a)
 			}
 		}
 	}
 	return apiutils.Deduplicate(allowed)
 }
 
+// WithAllowedKubernetesResourceKindFilter returns a SearchAsRolesOption func
+// that will check that the requestedKubeResourceKind exists in the allow list
+// for the current search as role.
+func WithAllowedKubernetesResourceKindFilter(requestedKubeResourceKind string, allowedLookup map[string][]string) SearchAsRolesOption {
+	return func(searchAsRoleName string) bool {
+		return slices.Contains(allowedLookup[searchAsRoleName], requestedKubeResourceKind)
+	}
+}
```

Comment:
imo building the lookup table, then creating this filer, to pass to GetAllowedSearchAsRoles, is more complicated than it needs to be

I think we need to deny the whole request if any of the user's base roles denies requestedKubeResourceKind, and I can't figure out if this does that or not

I think it would be simpler and clearer if you just made a method `func (set RoleSet) GetAllowedSearchAsRolesForKubeKind(requestedKubeResourceKind string) ([]string, error)`, iterate through the base roles, if any denies `requestedKubeResourceKind` return an error to deny the whole request. Add the search_as_role if the base role allows the requested kind. Should be pretty straightforward.

Reply:
sorry, i'm complicating things as usual ;;. 

i didn't return an error b/c didn't think it was necessary? if things get denied or there were no matches, it returns empty array?

Reply:
I agree with Lisa here.
Even if you can't request access to particular resources, one can always query the existing resources he can see, similarly to search teleport resources

Reply:
ah sorry you're right. I mean the user should not get any extra roles if any of their base roles explicitly denies requesting the kube resource kind, it doesn't need to be an error